### DISCUSSION TITLE
Package not scheduled, claims "not within schedule", despite seeming like it is within the schedule?

### DISCUSSION THREAD

**Discussion by lorengordon** [renovatebot/renovate#32369] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab, Renovate version 38.89.3

### Please tell us more about your question or problem

I have renovate running once a day, and a configuration in a repository that should create a pull request on the last Monday of the month. From the logs, the schedule is read correctly, and the date appears to me to be within the schedule, but renovate claims it is "not within schedule". I'm not sure what is happening, nor am I sure how to troubleshoot further. Any ideas?

### Logs (if relevant)

<details><summary>Logs</summary>

```
DEBUG: Checking schedule(schedule=* * * * * 1L, tz=America/New_York, now=2024-10-28T17:05:37.480Z) (repository=cirrus/cirrus-k8s, branch=renovate/all-terraform-dependencies)
DEBUG: Found timezone: America/New_York (repository=cirrus/cirrus-k8s, branch=renovate/all-terraform-dependencies)
DEBUG: Adjusting now for timezone (repository=cirrus/cirrus-k8s, branch=renovate/all-terraform-dependencies)
DEBUG: Checking 1 schedule(s) (repository=cirrus/cirrus-k8s, branch=renovate/all-terraform-dependencies)
DEBUG: Human-readable summary for cron:: Every second, on the last Monday of the month (repository=cirrus/cirrus-k8s, branch=renovate/all-terraform-dependencies)
DEBUG: Package not scheduled (repository=cirrus/cirrus-k8s, branch=renovate/all-terraform-dependencies)
DEBUG: Skipping branch creation as not within schedule (repository=cirrus/cirrus-k8s, branch=renovate/all-terraform-dependencies)
```

</details>

**Comment by rarkins** [comment]
This is the cronMatches function: https://github.com/renovatebot/renovate/blob/54bdb36c872042f4341edc66e49d101b3942c914/lib/workers/repository/update/branch/schedule.ts#L102

This is the test file for scheduling: https://github.com/renovatebot/renovate/blob/54bdb36c872042f4341edc66e49d101b3942c914/lib/workers/repository/update/branch/schedule.spec.ts

If there is a bug, I'm guessing it's the 'L' part. Can you reproduce the problem via a new unit test in that file? Otherwise we need to wait for the last week of the month to reproduce this.

**Reply by RahulGautamSingh** [reply]
It is the 'L', we do not support it yet.

**Reply by rarkins** [reply]
Does our cron parsing library support it / consider it valid? If we just need to add an extra "if" clause or two ourselves then it would be feasible to support, but not if it  means replacing our cron parser or something else large.

BTW I would prefer that we support "all" cron patterns with the only caveat being that minutes=*

**Reply by lorengordon** [reply]
Oh! I didn't realize that wasn't supported. It seemed like it was documented when I did it a month ago or so, but now I can't find where I got that syntax from. Mostly all I was trying to express is "Once a month on a weekday". Is there some other way to do that?

**Reply by RahulGautamSingh** [reply]
_For lorengordon_
1. It was documented I think, will add it again if not there.
2. `* * 1-7 * 0` (Once a month on a weekday) (from 1 to 7 each month but only on sunday)

_For Rhys_
3. Yes I think adding an if clause should do it

**Reply by lorengordon** [reply]
> * * 1-7 * 0 (Once a month on a weekday) (from 1 to 7 each month but only on sunday)

Is that handling done specifically in Renovate's cron parser? When I check the docs, it seems when day of month and day of week are specified, _both_ will match. For `* * 1-7 * 0`, that would mean it would run every day the first-seventh of the month, and also every Sunday... From the cron docs:

> Note: The day of a command's execution can be specified in the following two fields --- 'day of month', and 'day of week'. If both fields are restricted (i.e., do not contain the "*" character), the command will be run when either field matches the current time. For example,
"30 4 1,15 * 5" would cause a command to be run at 4:30 am on the 1st and 15th of each month, plus every Friday.
Is that
https://crontab.guru/crontab.5.html

**Reply by RahulGautamSingh** [reply]
> Is that handling done specifically in Renovate's cron parser?

Yes.

It is bit wrong to not follow the docs we recommend tbh. I will fix it up quick so that we allow `L` and `#1` for now ie. last and first

**Reply by lorengordon** [reply]
Fwiw, the update to `* * 1-7 * 0` did indeed trigger updates only once per month on the specified day of week. Really appreciate the fast response and guidance! Thanks so much folks!

**Reply by lorengordon** [reply]
Hmm,  @RahulGautamSingh, with the updates to cron support via https://github.com/renovatebot/renovate/issues/32414, was it changed so that `* * 1-7 * 0` now does what the cron docs say, and runs every day the first week of the month, and then also every Sunday? Instead of _just_ the first Sunday of the month, as it did previously with the custom cron handling? Just wanting to confirm, because that would explain why I've gotten merge requests every day so far this week, since I updated the renovate image... :D

**Reply by RahulGautamSingh** [reply]
Yes, the behavior switched from AND to OR in the PR you linked. But, we want to have the AND logic only and a fix to is on its way.

**Comment by rarkins** [comment]
Created https://github.com/renovatebot/renovate/issues/32414
